content/post/date_features.Rmd

```diff
@@ -0,0 +1,195 @@
+$\frac{2 \pi \mathbf{x}}{max_{\mathbf{x}}}$. This results in a two-dimensional representation of the
+feature:
+```{r, echo=FALSE, fig.height=5.5, fig.width=10, fig.align = "center"}
+hours = 0:23
+hours_scaled = (2 * pi * hours) / 24
+hours_sin = sin(hours_scaled)
+hours_cos = cos(hours_scaled)
+n = 24
+cols = rainbow(n, s = 1, v = 1, start = 0, end = max(1, n - 1) / n, alpha = 1)
+par(mfrow = c(1, 2))
+plot(hours, type = "n", xlab = "Index", ylab = "Hours")
+text(hours, col = cols, labels = as.character(hours))
+plot(hours_sin, hours_cos, type = "n", xlab = "Sine Transformation", ylab = "Cosine Transformation")
+text(hours_sin, hours_cos, col = cols, labels = as.character(hours))
+```
+
+`r mlr_pkg("mlr3pipelines")` provides the `PipeOpDateFeatures` pipeline which can be used to
+automatically engineer features based on `POSIXct` columns, including handling of cyclical features.
+This is useful as most learners naturally cannot handle dates and `POSIXct` variables and therefore
+require conversion prior to training.
+
+## Bike Sharing
+The [Bike Sharing Dataset](https://archive.ics.uci.edu/ml/datasets/bike+sharing+dataset) contains
+the hourly count of rental bikes between years 2011 and 2012 in Capital bikeshare system with the
+corresponding weather and seasonal information. The dataset can be downloaded from the UCI Machine
+Learning Repository. After reading in the data, we fix some factor levels, and convert some data
+types.
+```{r}
+tmp <- tempfile()
+download.file("https://archive.ics.uci.edu/ml/machine-learning-databases/00275/Bike-Sharing-Dataset.zip", tmp)
+bikes = read.csv(unz(tmp, filename = "hour.csv"), as.is = TRUE)
+bikes$season = factor(bikes$season, labels = c("winter", "spring", "summer", "fall"))
+bikes$holiday = as.logical(bikes$holiday)
+bikes$workingday = as.logical(bikes$workingday)
+bikes$weathersit = as.factor(bikes$weathersit)
+```
+
+Our goal will be to predict the total number of rented bikes on a given day: `cnt`.
+```{r}
+str(bikes)
+```
+
+The original dataset does not contain a `POSIXct` column, but we can easily generate one based on
+the other variables available (note that as no information regarding minutes and seconds is
+available, we set them to `:00:00`):
+```{r}
+bikes$date = as.POSIXct(strptime(paste0(bikes$dteday, " ", bikes$hr, ":00:00"), tz = "GMT",
+  format = "%Y-%m-%d %H:%M:%S"))
+summary(bikes$date)
+```
+
+## Baseline Random Forest
+We construct a new regression task and create a vector of train and test indices:
+```{r}
+library(mlr3)
+library(mlr3learners)
+set.seed(2906)
+tsk = TaskRegr$new("bikes", backend = bikes, target = "cnt")
+train.idx = sample(seq_len(tsk$nrow), size = 0.7 * tsk$nrow)
+test.idx = setdiff(seq_len(tsk$nrow), train.idx)
+```
+
+This allows us to construct a train and test task:
+```{r}
+tsk_train = tsk$clone()$filter(train.idx)
+tsk_test = tsk$clone()$filter(test.idx)
+```
+
+To estimate the performance on future data, we will use `3-fold cross-validation`:
```

Comment:
```suggestion
To estimate the performance on unseen data, we will use a `3-fold cross-validation`.
Note that this involves validating on past data, which is usually a bad practice but should suffice for this example.
```

Reply:
I agree that the cv is somewhat problematic. Thanks for the suggested changes! I guess this makes it more clear!